pkgdb/src/lock-flake-installable.cc

```diff
@@ -0,0 +1,320 @@
+
+  this->parser.add_argument( "url" )
+    .help( "The flake installable to lock" )
+    .action( [&]( const std::string & value ) { this->installable = value; } );
+
+  auto arg
+    = this->parser.add_argument( "--system" )
+        .metavar( "SYSTEM" )
+        .help( "The system to lock the flake installable for" )
+        .default_value( nix::settings.thisSystem.get() )
+        .nargs( 1 )
+        .action( [&]( const std::string & value ) { this->system = value; } );
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+int
+LockCommand::run()
+{
+  auto state = this->getState();
+
+  debugLog( nix::fmt( "original installable: %s", this->installable ) );
+
+
+  std::tuple<nix::FlakeRef, std::string_view, nix::ExtendedOutputsSpec> parsed
+    = nix::parseFlakeRefWithFragmentAndExtendedOutputsSpec( this->installable );
+
+  nix::FlakeRef            flakeRef            = std::get<0>( parsed );
+  std::string_view         fragment            = std::get<1>( parsed );
+  nix::ExtendedOutputsSpec extendedOutputsSpec = std::get<2>( parsed );
+
+  debugLog(
+    nix::fmt( "original flakeRef: '%s'", flakeRef.to_string().c_str() ) );
+  debugLog( nix::fmt( "original fragment: '%s'", fragment ) );
+  debugLog( nix::fmt( "original extendedOutputsSpec: '%s'",
+                      extendedOutputsSpec.to_string() ) );
+
+  auto lockFlags = nix::flake::LockFlags {
+    .recreateLockFile = false,
+    .updateLockFile   = false,
+    .writeLockFile    = false,
+    .allowUnlocked    = true,
+    .commitLockFile   = false,
+  };
+
+
+  auto installable = nix::make_ref<nix::InstallableFlake>(
+    static_cast<nix::SourceExprCommand *>( nullptr ),
+    state,
+    std::move( flakeRef ),
+    fragment,
+    extendedOutputsSpec,
+    nix::Strings {
+      "packages." + this->system + ".default",
+      "legacyPackages." + this->system + ".default",
+    },
+    nix::Strings {
+      "packages." + this->system + ".",
+      "legacyPackages." + this->system + ".",
+    },
+    lockFlags );
+  debugLog(
+    nix::fmt( "locked installable: '%s'", installable->what().c_str() ) );
+
+
+  auto lockedUrl = installable->getLockedFlake()->flake.lockedRef.to_string();
+  debugLog( nix::fmt( "locked url: '%s'", lockedUrl ) );
+
+  auto flakeDescription = installable->getLockedFlake()->flake.description;
+
+  auto cursor = installable->getCursor( *state );
+
+  auto lockedAttrPath = cursor->getAttrPathStr();
+  debugLog( nix::fmt( "locked attr path: '%s'", lockedAttrPath ) );
+
+  debugLog( nix::fmt( "locked outputs: '%s'",
+                      installable->extendedOutputsSpec.to_string() ) );
+
+  // check if the output is a derivation (not a just a store path)
+  if ( ! cursor->isDerivation() )
+    {
+      auto v = cursor->forceValue();
+      throw nix::Error(
+        "expected flake output attribute '%s' to be a derivation but found %s",
+        lockedAttrPath,
+        nix::showType( v ) );
```

Comment:
There are at least three different error cases:
1) incorrect flakeref
2) incorrect attrpath in flake (failing together with (1) while creating the `installable`)
3) attrpath is not a derivation (`type != "derivation")
4) attrset does not have a required attribute (unlikely for "well behaved" derivations, but possible in the global space)

Does it make sense to define separate exceptions for all these of are we fine with passing on the nix errors that may fall out of this?

Reply:
I at least defined an exception and mapped some obvious ones to it, but i'm not sure how much that actually affords us considering that only the derivation check stems from flox logic (which is also borrowed from upstream) and the rest are only wrap original nix errors with a bit of context:

e.g.:

```
Failed to lock flake installable: could not parse flake installable: input 'git+htts://github.com/nixos/nixpkgs#legacyPackages.aarch64-darwin.hello' is unsupported
```

instead of

```
error: input 'git+htts://github.com/nixos/nixpkgs#legacyPackages.aarch64-darwin.hello' is unsupported
```

To be complete we should probably catch all type casting errors for accessing nix attributes as well eventually.